From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: nds-ports <nds-ports@localhost>
Date: Wed, 7 Oct 2026 12:00:00 +0200
Subject: [PATCH] Speed up the drawers

The C versions of the assembly drawers kept everything in globals and
counted every pixel for a debugging feature. They keep their state in
registers now, the wall drawer does two columns at a time with a halfword
per row, and the pixel counter is gone. Drawers given no pixels to draw
draw none.
---
 Engine/src/draw.c   | 190 ++++++++++++++++----------------------------
 Engine/src/draw.h   |   1 -
 Engine/src/engine.c |   3 -
 3 files changed, 70 insertions(+), 124 deletions(-)

diff --git a/Engine/src/draw.c b/Engine/src/draw.c
index 00ca069..7210d35 100644
--- a/Engine/src/draw.c
+++ b/Engine/src/draw.c
@@ -10,8 +10,6 @@
 #include "build.h"
 #include "draw.h"
 
-int32_t pixelsAllowed = 10000000000;
-
 uint8_t  *transluc = NULL;
 
 static int transrev = 0;
@@ -56,34 +54,16 @@ void sethlinesizes(int32_t i1, int32_t _bits, uint8_t * textureAddress)
 //Draw a line from destination in the framebuffer to framebuffer-numPixels
 void hlineasm4(int32_t numPixels, int32_t shade, uint32_t i4, uint32_t i5, uint8_t *dest){
 
-    int32_t shifter = ((256-machxbits_al) & 0x1f);
-    uint32_t source;
-    
-    uint8_t * texture = textureSetup;
-    uint8_t bits = bitsSetup;
-    
-    shade = shade & 0xffffff00;
-    numPixels++;
-    
-	if (!RENDER_DRAW_CEILING_AND_FLOOR)
-		return;
-
-    while (numPixels) {
+    const uint8_t *texture = textureSetup;
+    const uint8_t *pal = globalpalwritten + (shade & 0xffffff00);
+    const int32_t shifter = ((256-machxbits_al) & 0x1f);
+    const int32_t bits = bitsSetup;
+    const uint32_t inc5 = asm1, inc4 = asm2;
 
-	    source = i5 >> shifter;
-	    source = shld(source,i4,bits);
-	    source = texture[source];
-        
-		if (pixelsAllowed-- > 0)
-			*dest = globalpalwritten[shade|source];
-        
-	    dest--;
-        
-	    i5 -= asm1;
-	    i4 -= asm2;
-        
-	    numPixels--;
-		
+    for (numPixels++; numPixels > 0; numPixels--) {
+        *dest-- = pal[texture[shld(i5 >> shifter, i4, bits)]];
+        i5 -= inc5;
+        i4 -= inc4;
     }
 }
 
@@ -128,7 +108,6 @@ void rhlineasm4(int32_t i1, uint8_t* texture, int32_t i3, uint32_t i4, uint32_t
 	    ebp &= rmach_esi;
 	    i1 = ((i1&0xffffff00)|(((uint8_t *)i3)[rmach_edx]));
 
-		if (pixelsAllowed-- > 0)
 			 ((uint8_t *)rmach6b)[numPixels] = (i1&0xff);
 
 	    texture -= ebp;
@@ -180,7 +159,6 @@ void rmhlineasm4(int32_t i1, intptr_t shade, int32_t colorIndex, int32_t i4, int
         
         //Check if this colorIndex is the transparent color (255).
 	    if ((colorIndex&0xff) != 255) {
-			if (pixelsAllowed-- > 0)
 			{
 				i1 = ((i1&0xffffff00)|(((uint8_t  *)colorIndex)[rmmach_edx]));
 				((uint8_t  *)rmach6b)[numPixels] = (i1&0xff);
@@ -220,7 +198,6 @@ int32_t prevlineasm1(int32_t i1, uint8_t* palette, int32_t i3, int32_t i4, uint8
         i4 = ((uint32_t)i4) >> mach3_al;
 	    i4 = (i4&0xffffff00) | source[i4];
 
-		if (pixelsAllowed-- > 0)
 			*dest = palette[i4];
 
 		
@@ -235,26 +212,15 @@ int32_t prevlineasm1(int32_t i1, uint8_t* palette, int32_t i3, int32_t i4, uint8
 //FCS: This is used to draw wall border vertical lines
 int32_t vlineasm1(int32_t vince, uint8_t* palookupoffse, int32_t numPixels, int32_t vplce, uint8_t* texture, uint8_t* dest)
 {
-    uint32_t temp;
-
-    if (!RENDER_DRAW_WALL_BORDERS)
-		return vplce;
+    const int32_t shift = mach3_al, pitch = bytesperline;
+    uint32_t v = vplce;
 
-    numPixels++;
-    while (numPixels)
-    {
-	    temp = ((uint32_t)vplce) >> mach3_al;
-        
-	    temp = texture[temp];
-      
-		if (pixelsAllowed-- > 0)
-			*dest = palookupoffse[temp];
-	    
-		vplce += vince;
-	    dest += bytesperline;
-	    numPixels--;
+    for (numPixels++; numPixels > 0; numPixels--) {
+        *dest = palookupoffse[texture[v >> shift]];
+        v += vince;
+        dest += pitch;
     }
-    return vplce;
+    return v;
 } 
 
 
@@ -280,7 +246,6 @@ int32_t tvlineasm1(int32_t i1, uint8_t  * texture, int32_t numPixels, int32_t i4
 			if (transrev) 
 				colorIndex = ((colorIndex>>8)|(colorIndex<<8));
             
-			if (pixelsAllowed-- > 0)
 				*dest = transluc[colorIndex];
 		}
         
@@ -334,7 +299,6 @@ void tvlineasm2(uint32_t i1, uint32_t i2, uintptr_t i3, uintptr_t i4, uint32_t i
 				if (transrev) 
 					val = ((val>>8)|(val<<8));
 
-				if (pixelsAllowed-- > 0)
 					((uint8_t  *)i6)[tran2edi1] = transluc[val];
 			}
 		} else if (i4 == 255) { // skipdraw2
@@ -345,7 +309,6 @@ void tvlineasm2(uint32_t i1, uint32_t i2, uintptr_t i3, uintptr_t i4, uint32_t i
 			if (transrev) 
                 val = ((val>>8)|(val<<8));
 
-			if (pixelsAllowed-- > 0)
 				((uint8_t  *)i6)[tran2edi] = transluc[val];
 		} else {
 			uint16_t l = ((uint8_t  *)i6)[tran2edi]<<8;
@@ -356,11 +319,9 @@ void tvlineasm2(uint32_t i1, uint32_t i2, uintptr_t i3, uintptr_t i4, uint32_t i
 				l = ((l>>8)|(l<<8));
 				r = ((r>>8)|(r<<8));
 			}
-			if (pixelsAllowed-- > 0)
 			{
 				((uint8_t  *)i6)[tran2edi] = transluc[l];
 				((uint8_t  *)i6)[tran2edi1] =transluc[r];
-				pixelsAllowed--;
 			}
 		}
 		i6 += bytesperline;
@@ -374,23 +335,19 @@ void tvlineasm2(uint32_t i1, uint32_t i2, uintptr_t i3, uintptr_t i4, uint32_t i
 static uint8_t  machmv;
 int32_t mvlineasm1(int32_t vince, uint8_t* palookupoffse, int32_t i3, int32_t vplce, uint8_t* texture, uint8_t  *dest)
 {
-    uint32_t temp;
+    const int32_t shift = machmv, pitch = bytesperline;
+    uint32_t v = vplce;
 
     for(;i3>=0;i3--)
     {
-		temp = ((uint32_t)vplce) >> machmv;
-	    temp = texture[temp];
+        uint32_t temp = texture[v >> shift];
 
-	    if (temp != 255) 
-		{
-			if (pixelsAllowed-- > 0)
-			*dest = palookupoffse[temp];
-		}
-
-	    vplce += vince;
-	    dest += bytesperline;
+        if (temp != 255)
+            *dest = palookupoffse[temp];
+        v += vince;
+        dest += pitch;
     }
-    return vplce;
+    return v;
 }
 
 
@@ -400,36 +357,39 @@ void setupvlineasm(int32_t i1)
 }
 
 //FCS This is used to fill the inside of a wall (so it draws VERTICAL column, always).
-void vlineasm4(int32_t columnIndex, intptr_t framebuffer)
+//Two columns at a time, a halfword per row, so everything fits in registers
+static inline void vlineasm2(int32_t numRows, uint16_t *dest, int32_t i)
 {
+    const int32_t shift = mach3_al, pitch = bytesperline >> 1;
+    const int32_t i0 = vince[i], i1 = vince[i + 1];
+    const uint8_t *b0 = (const uint8_t *)bufplce[i], *b1 = (const uint8_t *)bufplce[i + 1];
+    const uint8_t *p0 = palookupoffse[i], *p1 = palookupoffse[i + 1];
+    uint32_t v0 = vplce[i], v1 = vplce[i + 1];
 
-	if (!RENDER_DRAW_WALL_INSIDE)
-		return ;
+    do {
+        *dest = p0[b0[v0 >> shift]] | (p1[b1[v1 >> shift]] << 8);
+        v0 += i0;
+        v1 += i1;
+        dest += pitch;
+    } while (--numRows);
+    vplce[i] = v0;
+    vplce[i + 1] = v1;
+}
 
+void vlineasm4(int32_t numRows, intptr_t framebuffer)
+{
+    int i;
+
+    // Walls are halfword aligned, sprites drawn by rotatesprite may not be
+    if (((framebuffer | bytesperline) & 1) == 0)
     {
-        int i;
-        uint32_t temp;
-        
-        uintptr_t index = (framebuffer + ylookup[columnIndex]);
-        uint8_t  *dest= (uint8_t *)(-ylookup[columnIndex]);
-        //uint8_t  *dest= (uint8_t *)framebuffer;
-        
-        //uint32_t index = 0;
-        do {
-            for (i = 0; i < 4; i++)
-            {
-				
-        	    temp = ((uint32_t)vplce[i]) >> mach3_al;
-        	    temp = (((uint8_t *)(bufplce[i]))[temp]);
-                
-				if (pixelsAllowed-- > 0)
-        			dest[index+i] = palookupoffse [i] [temp];
-                
-	            vplce[i] += vince[i];
-            }
-            dest += bytesperline;
-        } while (((uint32_t)dest - bytesperline) < ((uint32_t)dest));
+        vlineasm2(numRows, (uint16_t *)framebuffer, 0);
+        vlineasm2(numRows, (uint16_t *)framebuffer + 1, 2);
+        return;
     }
+    for (i = 0; i < 4; i++)
+        vplce[i] = vlineasm1(vince[i], palookupoffse[i], numRows - 1, vplce[i],
+                             (uint8_t *)bufplce[i], (uint8_t *)framebuffer + i);
 } 
 
 
@@ -440,33 +400,29 @@ void setupmvlineasm(int32_t i1)
 } 
 
 
-void mvlineasm4(int32_t column, intptr_t framebufferOffset)
+//Four columns of a masked wall, 255 is transparent
+void mvlineasm4(int32_t numRows, intptr_t framebufferOffset)
 {
     int i;
-    uint32_t temp;
-    uintptr_t index = (framebufferOffset + ylookup[column]);
-    uint8_t  *dest = (uint8_t *)(-ylookup[column]);
 
-    do {
-
-		if (pixelsAllowed <= 0)
-			return;
+    for (i = 0; i < 4; i++)
+    {
+        const int32_t shift = machmv, inc = vince[i];
+        const uint8_t *texture = (const uint8_t *)bufplce[i], *pal = palookupoffse[i];
+        uint32_t v = vplce[i];
+        uint8_t *dest = (uint8_t *)framebufferOffset + i;
+        int32_t n = numRows;
 
-        for (i = 0; i < 4; i++)
-        {
-			
-	      temp = ((uint32_t)vplce[i]) >> machmv;
-	      temp = (((uint8_t *)(bufplce[i]))[temp]);
-	      if (temp != 255)
-		  {
-			  if (pixelsAllowed-- > 0)
-				dest[index+i] = palookupoffse[i][temp];
-		  }
-	      vplce[i] += vince[i];
-        }
-        dest += bytesperline;
+        do {
+            uint32_t temp = texture[v >> shift];
 
-    } while (((uint32_t)dest - bytesperline) < ((uint32_t)dest));
+            if (temp != 255)
+                *dest = pal[temp];
+            v += inc;
+            dest += bytesperline;
+        } while (--n);
+        vplce[i] = v;
+    }
 } 
 /* END ---------------  WALLS RENDERING METHOD (USED TO BE HIGHLY OPTIMIZED ASSEMBLY) ----------------------------*/
 
@@ -531,7 +487,6 @@ setup:
         
         i1 = (i1&0xffffff00) | (((uint8_t  *)spal_eax)[i1]&0xff);
         
-        if (pixelsAllowed-- > 0)
             *dest = i1;
         
         dest += bytesperline;
@@ -591,7 +546,6 @@ setup:
         {
             colorIndex = (colorIndex&0xffffff00) | (((uint8_t  *)spal_eax)[colorIndex]&0xff);
             
-            if (pixelsAllowed-- > 0)
                 *dest = colorIndex;
         }
    
@@ -671,7 +625,6 @@ void DrawSpriteVerticalLine(int32_t i2, int32_t numPixels, uint32_t i4, uint8_t
 
 				colorIndex = transluc[val];
 
-				if (pixelsAllowed-- > 0)
 					*dest = colorIndex;
 			}
             
@@ -743,7 +696,6 @@ void mhlineskipmodify( uint32_t i2, int32_t numPixels, int32_t i5, uint8_t* dest
 
         //Skip transparent color.
 		if ((colorIndex&0xff) != 0xff){
-            if (pixelsAllowed-- > 0)
 				*dest = mmach_asm3[colorIndex];
         }
 	    i2 += mmach_asm1;
@@ -797,7 +749,6 @@ void thlineskipmodify(int32_t i1, uint32_t i2, uint32_t i3, int32_t i4, int32_t
 		    if (transrev) 
 				val = ((val>>8)|(val<<8));
 
-			if (pixelsAllowed-- > 0)
 			 *i6 = transluc[val];
 	    }
 
@@ -903,7 +854,6 @@ void slopevlin(intptr_t i1, uint32_t i2, int32_t i3, int32_t i4, int32_t i5, int
 		    eax = ((eax&0xffffff00)|(*((uint8_t  *)(ebx+edx))));
 		    ebx = esi;
 
-			if (pixelsAllowed-- > 0)
 				*((uint8_t  *)i1) = (eax&0xff);
 
 		    edx = edi;
diff --git a/Engine/src/draw.h b/Engine/src/draw.h
index 0f2a656..daaad3f 100644
--- a/Engine/src/draw.h
+++ b/Engine/src/draw.h
@@ -81,7 +81,6 @@ void settrans(int32_t type);
 //VISUALIZE RENDERER
 
 #define MAX_PIXEL_RENDERERED ((1600*1200)+20000 )
-extern int32_t pixelsAllowed;
 
 #define RENDER_DRAW_WALL_BORDERS 1
 #define RENDER_DRAW_WALL_INSIDE 1
diff --git a/Engine/src/engine.c b/Engine/src/engine.c
index b910987..dbafb5a 100644
--- a/Engine/src/engine.c
+++ b/Engine/src/engine.c
@@ -2807,9 +2807,6 @@ void drawrooms(int32_t daposx, int32_t daposy, int32_t daposz,short daang, int32
 	if (pixelRenderable >= MAX_PIXEL_RENDERERED)
 		pixelRenderable =  0 ;
 
-	//pixelsAllowed = pixelRenderable;
-	pixelsAllowed = 100000000;
-	//printf("%d\n",pixelsAllowed);
 
     beforedrawrooms = 0;
 
